pravega/pravega-operator
# Issue 96: Startup and healthcheck improvements

## Change log description

#### Readiness probes
[Readiness probes](https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-probes/#define-readiness-probes) have been added to prevent the pods from becoming _ready_ until they have started up correctly and are actually accepting requests. 

Bookkeeper and Controller pods should start quickly because they only depend on ZooKeeper, which should be already up and ready. The maximum time allowed for them to become ready is 1.5 and 1 minute respectively.

Segment Store pods depend on BookKeeper, and if configured with external access enabled, they will have to wait for the service external IP to be allocated. In GKE, we have experienced up to three minutes for the external IP to be allocated and obtained by the pod, therefore, the timeout to become ready has been set to 5 minutes. 

The command to verify pod readiness is the same for all components and checks whether or not there's a process listening on the relevant port.

#### Liveness probe
In addition to the Readiness probes, [Liveness probes](https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-probes/#define-a-liveness-command) have been added to detect and remedy situations in which the pod transitions to a broken state. If the pod becomes unhealthy, the liveness probe will detect it and restart it. The command to verify pod liveness is the same used for readiness.

#### Segment Store wait for Bookkeeper on startup
Starting Segment Store and BookKeeper pods at the same time was causing Segment Store instances to throw errors and restart a couple of times until Bookies were up and accepting requests. 

Pravega Docker images accept a [`WAIT_FOR` environment variable](https://github.com/pravega/pravega/blob/master/docker/compose/docker-compose.yml#L89) which tells the container to wait for one or more systems to become reachable until the container is allowed to start. This PR leverages that option to configure Segment Store pods to wait until at least three Bookies become ready.

#### Segment Store sequential, ordered startup
To avoid errors in the bootstrap process, Segment Store pods will now start sequentially, one after another. This is also aligned with how external IP are sequentially allocated and with the new readiness and liveness probes, i.e., if we were to start all Segment Store in parallel, at the same time, the pods with higher rank would time out their readiness probes before their external IP addresses are allocated. 

## Purpose of the change
Fix #96 

## How to test the change
Checkout the repository (`master` branch is fine, no need to checkout this branch).

Edit the operator image used in the `deploy/operator.yaml` file to use the `adrianmo/pravega-operator:issue-96-6` image, which contains the changes in this PR.

Deploy the operator.

```
kubectl create -f deploy/operator.yaml
```

Follow the instructions in the README file to deploy ZooKeeper, NFS provisioner, and a Pravega cluster with 3 Bookies, 1 or more Controllers, and 3 Segment stores.


When you deploy a Pravega cluster, check that Segment Store instances are created sequentially.

```
$ kubectl get po -l pravega_cluster=pravega
NAME                                          READY     STATUS    RESTARTS   AGE
pravega-bookie-0                              1/1       Running   0          44s
pravega-bookie-1                              1/1       Running   0          44s
pravega-bookie-2                              1/1       Running   0          44s
pravega-pravega-controller-5c7556c797-x5694   1/1       Running   0          44s
pravega-pravega-segmentstore-0                0/1       Running   0          44s
```

Check the logs of Segment Store 0 pod to see how it tries to connect to the Bookies.

```
$ kubectl logs pravega-pravega-segmentstore-0
Waiting for pravega-bookie-0.pravega-bookie-headless.default:3181
Unable to connect to pravega-bookie-0.pravega-bookie-headless.default:3181. Waiting
[Errno -2] Name does not resolve
Unable to connect to pravega-bookie-0.pravega-bookie-headless.default:3181. Waiting
[Errno -2] Name does not resolve
Unable to connect to pravega-bookie-0.pravega-bookie-headless.default:3181. Waiting
[Errno -2] Name does not resolve
Connected to pravega-bookie-0.pravega-bookie-headless.default:3181
Waiting for pravega-bookie-1.pravega-bookie-headless.default:3181
Unable to connect to pravega-bookie-1.pravega-bookie-headless.default:3181. Waiting
[Errno -2] Name does not resolve
Connected to pravega-bookie-1.pravega-bookie-headless.default:3181
Waiting for pravega-bookie-2.pravega-bookie-headless.default:3181
Connected to pravega-bookie-2.pravega-bookie-headless.default:3181
...
```

As soon as `pravega-pravega-segmentstore-0` becomes _ready_, `pravega-pravega-segmentstore-1` should be created.

```
$ kubectl get po -l pravega_cluster=pravega
NAME                                          READY     STATUS    RESTARTS   AGE
pravega-bookie-0                              1/1       Running   0          3m
pravega-bookie-1                              1/1       Running   0          3m
pravega-bookie-2                              1/1       Running   0          3m
pravega-pravega-controller-5c7556c797-x5694   1/1       Running   0          3m
pravega-pravega-segmentstore-0                1/1       Running   0          3m
pravega-pravega-segmentstore-1                0/1       Running   0          20s
```

Inspect all pods to check that Liveness and Readiness have been set successfully.

```
$ kubectl describe pod pravega-bookie-1 |grep 'Liveness\|Readiness'
    Liveness:   exec [/bin/sh -c netstat -ltn | grep 3181 || ss -ltn | grep 3181] delay=60s timeout=1s period=15s #success=1 #failure=4
    Readiness:  exec [/bin/sh -c netstat -ltn | grep 3181 || ss -ltn | grep 3181] delay=0s timeout=1s period=10s #success=1 #failure=9
```
